autoprotocol/protocol.py

```diff
@@ -5046,85 +5046,109 @@ def provision(self, resource_id, dests, volumes):
           one should be specified explicitly in a list matching the order of the
           specified destinations.
+          Note: If value is provided for this field, value for 'amounts' must be skipped
 
         Raises
         ------
         TypeError
             If resource_id is not a string.
         RuntimeError
-            If length of the list of volumes specified does not match the number
+            If length of the list of volumes or masses specified does not match the number
             of destination wells specified.
         TypeError
-            If volume is not specified as a string or Unit (or a list of either)
+            If the unit provided is not supported
+            If volume or mass is not specified as a string or Unit (or a list of either)
         ValueError
-            Volume to provision exceeds max capacity of well
+            If the resource measurement mode is volume and the provision exceeds max capacity of well.
+            If the provisioning of volumes or amounts are not supported.
 
         Returns
         -------
         list(Provision)
             :py:class:`autoprotocol.instruction.Provision` instruction object(s) to be appended and returned
 
         """
+
         # Check valid well inputs
         if not is_valid_well(dests):
             raise TypeError(
                 "Destinations (dests) must be of type Well, "
                 "list of Wells, or WellGroup."
             )
         dests = WellGroup(dests)
+
         if not isinstance(resource_id, str):
             raise TypeError("Resource ID must be a string.")
-        if not isinstance(volumes, list):
-            volumes = [Unit(volumes)] * len(dests)
+
+        if (volumes is None and amounts is None) or (volumes and amounts):
+            raise ValueError(
+                "Either volumes or amounts should have value/s, but not both."
+            )
+
+        if volumes:
+            provision_amounts = volumes
         else:
-            if len(volumes) != len(dests):
+            provision_amounts = amounts
+
+        if not isinstance(provision_amounts, list):
+            provision_amounts = [Unit(provision_amounts)] * len(dests)
+        else:
+            if len(provision_amounts) != len(dests):
                 raise RuntimeError(
                     "To provision a resource into multiple "
-                    "destinations with multiple volumes, the  "
-                    "list of volumes must correspond with the "
+                    "destinations with multiple volumes or masses, the  "
+                    "list of volumes or masses must correspond with the "
                     "destinations in length and in order."
                 )
-            volumes = [Unit(v) for v in volumes]
-        for v in volumes:
-            if not isinstance(v, (str, Unit)):
-                raise TypeError("Volume must be a string or Unit.")
+            provision_amounts = [Unit(v) for v in provision_amounts]
```

Comment:
missed this from the previous review, but fwiw, you can do `[parse_unit(v, ["liter", "gram") for v in provision_amounts]` instead of `[Unit(v) for v in provision_amounts]`

That way, there'll be additional type checks earlier on. Its not expensive and wraps it with nicer error messages

Reply:
Fixed